File: core/src/main/java/com/geometryduel/neat/BehaviorSignature.java
package com.geometryduel.neat;

/**
 * 从 MatchStats 提取的行为特征向量，用于新颖性搜索 (C)。
 * 维度 ~14，归一化到 [0,1]。
 */
public class BehaviorSignature {
    public static final int DIM = 12;
    public final float[] vec = new float[DIM];

    /** 从对战统计构建行为签名 */
    public static BehaviorSignature from(MatchStats m, Genome g) {
        BehaviorSignature s = new BehaviorSignature();
        int play = Math.max(1, m.frames - MatchStats.COUNTDOWN_FRAMES);
        float pf = play;

        s.vec[0]  = clamp(m.hitsDealt / 10f);                // 命中数
        s.vec[1]  = clamp(m.hitsTaken / 10f);                // 受击数
        s.vec[2]  = clamp(m.shotsFired / 60f);               // 短弓频率
        s.vec[3]  = clamp(m.longShotsFired / 15f);           // 长弓频率
        s.vec[4]  = clamp(m.teleportsUsed / 10f);            // 传送频率
        s.vec[5]  = clamp(m.campFrames / pf);                // 墙壁时间比
        s.vec[6]  = clamp(m.centerFrames / pf);              // 中心时间比
        s.vec[7]  = clamp(m.activeMoveFrames / pf);          // 主动移动比
        s.vec[8]  = clamp(m.teleportKills / 5f);             // 传送击杀
        s.vec[9]  = clamp(m.longShotsHit / 5f);              // 长弓命中
        s.vec[10] = g != null ? clamp(g.nodes.size() / 60f) : 0f;  // 网络规模
        s.vec[11] = g != null ? clamp(g.conns.size() / 400f) : 0f; // 连接数
        return s;
    }

    /** 与另一个签名的欧氏距离 */
    public float distance(BehaviorSignature o) {
        float sum = 0f;
        for (int i = 0; i < DIM; i++) {
            float d = vec[i] - o.vec[i];
            sum += d * d;
        }
        return (float) Math.sqrt(sum);
    }

    private static float clamp(float v) { return v < 0f ? 0f : (v > 1f ? 1f : v); }
}

File: app/src/main/java/com/geometryduel/neat/BehaviorSignature.kt
package com.geometryduel.neat

import kotlin.math.sqrt

/**
 * 从 MatchStats 提取的行为特征向量，用于新颖性搜索 (C)。
 * 维度 12，归一化到 [0,1]。
 */
class BehaviorSignature {

    companion object {
        const val DIM = 12

        fun from(m: MatchStats, g: Genome?): BehaviorSignature {
            val s = BehaviorSignature()
            s.fill(m, g)
            return s
        }
    }

    val vec = FloatArray(DIM)

    /** 复用对象填充 */
    fun fill(m: MatchStats, g: Genome?) {
        val play = maxOf(1, m.frames - MatchStats.COUNTDOWN_FRAMES)
        val pf = play.toFloat()
        vec[0] = clamp(m.hitsDealt / 10f)
        vec[1] = clamp(m.hitsTaken / 10f)
        vec[2] = clamp(m.shotsFired / 60f)
        vec[3] = clamp(m.longShotsFired / 15f)
        vec[4] = clamp(m.teleportsUsed / 10f)
        vec[5] = clamp(m.campFrames / pf)
        vec[6] = clamp(m.centerFrames / pf)
        vec[7] = clamp(m.activeMoveFrames / pf)
        vec[8] = clamp(m.teleportKills / 5f)
        vec[9] = clamp(m.longShotsHit / 5f)
        vec[10] = if (g != null) clamp(g.nodes.size / 60f) else 0f
        vec[11] = if (g != null) clamp(g.conns.size / 400f) else 0f
    }

    /** 与另一个签名的欧氏距离 */
    fun distance(o: BehaviorSignature): Float {
        var sum = 0f
        for (i in 0 until DIM) {
            val d = vec[i] - o.vec[i]
            sum += d * d
        }
        return sqrt(sum)
    }

    /**
     * 行为策略档案：关键行为维度离散化为 5 位三档编码（共 243 种），
     * 供训练器统计稀有度。基于真实对局行为，而非适应度数值。
     */
    fun profile(): String {
        val sb = StringBuilder(5)
        sb.append(bin(vec[0]))  // 攻击性（命中）
        sb.append(bin(vec[3]))  // 长弓倾向
        sb.append(bin(vec[4]))  // 传送机动
        sb.append(bin(vec[5]))  // 蹲坑程度
        sb.append(bin(vec[7]))  // 主动移动
        return sb.toString()
    }

    private fun bin(v: Float): Char = if (v < 0.25f) '0' else if (v < 0.6f) '1' else '2'

    private fun clamp(v: Float) = if (v < 0f) 0f else if (v > 1f) 1f else v
}
